Nouvelle fonctionnalité – Ajout des claps :clap:

PetitMote opened the pull request:
Ajout des claps, système *à la medium* pour récompenser un contenu. Voir #6094 
closes #6094 

Actuellement, avec l'aide d'Amaury, on a :
- Créé le modèle en BDD
- Créé l'API pour ajouter des claps

Spécifications :
- [x] On peut ajouter plusieurs claps par contenu, dans une limite fixée en conf (actuellement à 100 claps)
- [x] Les utilisateurs sont identifié s'ils sont connectés, sinon on effectue un hash de l'ip, salé par la clef Django
- [x] Les claps sont datés
- [ ] Mettre en place des tests unitaires
- [ ] À la requête d'un contenu, nombre actuel de claps sur ce contenu
- [ ] À la requête d'un contenu, nombre de claps actuellement donnés par l'utilisateur
- [ ] À la requête d'un contenu, limite de claps
- [ ] Mettre en place des tests unitaires

Partie front :
- [ ] Affichage du nombre de claps
- [ ] Affichage des claps de l'utilisateur
- [ ] Mettre en place des tests unitaires
- [ ] Bouton de clap
- [ ] Requête de clap, avec bufferisation en local avant envoi des claps
- [ ] Gère la limite de claps dans la fonction en front
- [ ] Mettre en place des tests unitaires

À terme :
- Statistiques datées pour les auteurs de contenu
- Valeur par utilisateur (clap moyen)

La PR nécessite la mise à jour de la base de données.


### Contrôle qualité

Actuellement, il est possible de faire la migration avec `python mange.py migrate`. Il est possible de tester le fonctionnement de l'API. La PR étant en cours, cette partie sera mise à jour plus tard.

PetitMote committed:
Modèle de clap en BDD

PetitMote committed:
Modifications du modèle de clap
Notamment ajout de la fonction de hash ip

Co-authored-by: AmauryCarrad <amaury@carrade.eu>

PetitMote committed:
Lien API vers Claps

Co-authored-by: AmauryCarrad <amaury@carrade.eu>

PetitMote committed:
Ajout de l'API de claps

Co-authored-by: AmauryCarrad <amaury@carrade.eu>

PetitMote committed:
Ajout réglage claps max par contenu par personne

Co-authored-by: AmauryCarrad <amaury@carrade.eu>

PetitMote committed:
Migrations BDD pour claps

Co-authored-by: AmauryCarrade <amaury@carrade.eu>

PetitMote commented:
Je me rends compte que j'ai zappé de modifier ça, dans le modèle le hash_ip_address est limité à 40 caractères. S'agissant d'un SHA256, je ne sais pas quelle taille il fait en sortie. Sur les tests, on n'a pas eu d'erreur, mais je ne sais pas comment ça se comporte en réel.

Situphen commented on zds/tutorialv2/models/database.py:
```diff
@@ -1515,4 +1516,27 @@ def transfer_paternity_receiver(sender, instance, **kwargs):
+    @staticmethod
+    def hash_ip(request) -> str:
+        ip_address = request.META.get("HTTP_X_FORWARDED_FOR") or request.META.get("REMOTE_ADDR")
+        return str(hashlib.sha256(bytes(settings.SECRET_KEY + ip_address, "utf-8")).hexdigest())
```
`hexdigest` retourne déjà un objet `str` :)
```suggestion
        return hashlib.sha256(bytes(settings.SECRET_KEY + ip_address, "utf-8")).hexdigest()
```

Situphen commented on zds/tutorialv2/models/database.py:
```diff
@@ -1515,4 +1516,27 @@ def transfer_paternity_receiver(sender, instance, **kwargs):
+        db_index=True,
+    )
+    user = models.ForeignKey(User, db_index=True, on_delete=models.SET_NULL, null=True)
+    hash_ip_address = models.CharField(db_index=True, max_length=40)
```
`hash_ip` retourne une chaîne de 64 caractères maximum
```suggestion
    hash_ip_address = models.CharField(db_index=True, max_length=64)
```

PetitMote replied:
Alors du coup, ça se passe comment pour les migrations ? Si je les regénère classiquement, ça va faire 2 fichiers de migrations, non ?

Situphen replied:
Oui, le mieux est probablement de supprimer le fichier de migration actuel et de refaire `python manage.py makemigrations`.

PetitMote replied:
D'accord, merci. Et ça ne risque pas de coincer avec la BDD que j'ai en l'état ?

Situphen replied:
Peut-être oui, auquel cas il faudra lancer `make new-db` pour en avoir une neuve.

Situphen commented on zds/tutorialv2/api/views.py:
```diff
@@ -153,3 +156,33 @@ def get_queryset(self):
         return exports
+
+
+class ClapView(APIView):
+    def post(self, request, *args, **kwargs):
+        try:
+            claps_count = int(request.data["claps_count"])
+
+            if claps_count < 0:
+                raise ValueError
+        except (ValueError, KeyError):
+            return Response(None, status=status.HTTP_400_BAD_REQUEST)
+
+        content = get_object_or_404(PublishableContent.objects, pk=kwargs["pk"])
+        user = request.user if request.user.is_authenticated else None
+        hash_ip_address = Clap.hash_ip(request) if request.user.is_anonymous else ""
+
+        claps_by_user = (
+            Clap.objects.filter(content=content, user=user, hash_ip_address=hash_ip_address).aggregate(
+                Sum("claps_count")
+            )["claps_count__sum"]
+        ) or 0
+
+        if claps_by_user + claps_count > settings.ZDS_APP["content"]["max_claps_per_content"]:
+            claps_count = settings.ZDS_APP["content"]["max_claps_per_content"] - claps_by_user
+
+        if claps_count > 0:
+            clap = Clap(content=content, user=user, hash_ip_address=hash_ip_address, claps_count=claps_count)
+            clap.save()
+
+        return Response({"claps": claps_by_user + claps_count}, status=status.HTTP_201_CREATED)
```
Je propose de renommer certaines variables pour que le code soit mieux compréhensible. Je ne sais pas s'il serait bien de renommer aussi `request.data["claps_count"]` et `Clap.claps_count`.
```suggestion
class ClapView(APIView):
    def post(self, request, *args, **kwargs):
        try:
            additional_claps = int(request.data["claps_count"])

            if additional_claps < 0:
                raise ValueError
        except (ValueError, KeyError):
            return Response(None, status=status.HTTP_400_BAD_REQUEST)

        content = get_object_or_404(PublishableContent.objects, pk=kwargs["pk"])
        user = request.user if request.user.is_authenticated else None
        hash_ip_address = Clap.hash_ip(request) if request.user.is_anonymous else ""

        current_claps = (
            Clap.objects.filter(content=content, user=user, hash_ip_address=hash_ip_address).aggregate(
                Sum("claps_count")
            )["claps_count__sum"]
        ) or 0

        if current_claps + additional_claps > settings.ZDS_APP["content"]["max_claps_per_content"]:
            additional_claps = settings.ZDS_APP["content"]["max_claps_per_content"] - current_claps

        if additional_claps > 0:
            clap = Clap(content=content, user=user, hash_ip_address=hash_ip_address, claps_count=additional_claps)
            clap.save()

        return Response({"claps": current_claps + additional_claps}, status=status.HTTP_201_CREATED)
```

AmauryCarrade replied:
Je serais partisan de renommer les `claps_count` en `claps`, mais Moté s'y opposait

PetitMote replied:
Dans cette fonction, `claps_count` devait à l'origine correspondre à un `claps_count` dans l'objet `Clap`, mais vu qu'on a changé le traitement, c'est vrai que ça ne colle plus du coup.

Je suis assez favorable aux 2 propositions, sauf si on trouve encore mieux avant 😄 

Concernant le `claps_count`, je pense toujours la même chose, mais si plus de gens pensent que juste `claps` suffit, pas de soucis.

PetitMote committed:
Hexidgest already return a string

Co-authored-by: Situphen <Situphen@users.noreply.github.com>

firm1 commented on zds/tutorialv2/models/database.py:
```diff
@@ -1515,4 +1516,27 @@ def transfer_paternity_receiver(sender, instance, **kwargs):
+
+    @staticmethod
+    def hash_ip(request) -> str:
+        ip_address = request.META.get("HTTP_X_FORWARDED_FOR") or request.META.get("REMOTE_ADDR")
```
Tu devrais te servir de la fonction [get_client_ip](https://github.com/zestedesavoir/zds-site/blob/b57d3e6b21938ec7c5d6eb0a4e40dcc56a904f8f/zds/member/views.py#L1359) plutôt que de la réecrire.

PetitMote replied:
Merci !
(tu me donnerais ton mail Github que je te co-author dessus ?)

firm1 replied:
pas besoin de me mettre co-auteur pour si peu.

firm1 commented on zds/tutorialv2/api/views.py:
```diff
@@ -153,3 +156,33 @@ def get_queryset(self):
+            )["claps_count__sum"]
+        ) or 0
+
+        if claps_by_user + claps_count > settings.ZDS_APP["content"]["max_claps_per_content"]:
```
Pour simplifier ton `if` je te propose de le remplacer par:
```python
claps_count = min(settings.ZDS_APP["content"]["max_claps_per_content"], claps_by_user + claps_count)

PetitMote replied:
Tu proposes donc de remplacer :
```Python
if current_claps + additional_claps > settings.ZDS_APP["content"]["max_claps_per_content"]:
    additional_claps = settings.ZDS_APP["content"]["max_claps_per_content"] - current_claps
```

Par :
```Python
claps_count = min(settings.ZDS_APP["content"]["max_claps_per_content"], claps_by_user + claps_count)
claps_count = claps_count - claps_by_user
```

Si je comprends bien ?

firm1 replied:
Oublie ce commentaire, j'avais mal compris le code d'origine.

PetitMote committed:
Change taille max du hash ip

Co-authored-by: Situphen <Situphen@users.noreply.github.com>

PetitMote committed:
Change des noms de variables (clarté)

Co-authored-by: Situphen <Situphen@users.noreply.github.com>

PetitMote committed:
Utilisation d'une fonction get_client_ip existante

Co-authored-by: firm1

PetitMote committed:
Migration de la BDD pour les claps

PetitMote committed:
Changé la fonction hash_ip_address à cause d'une référence circulaire

La fonction appelait une autre fonction get_client_ip, et avait besoin de l'importer dans `zds.tutorialv2.models.database.py`. Cela créait une référence circulaire. Pour corriger le problème, la fonction hash_ip_address prend maintenant en paramètre directement l'address ip, et l'appel à get_client_ip se fait maintenant dans la view de l'API.

AmauryCarrade commented on zds/tutorialv2/migrations/0032_add_claps.py:
```diff
@@ -17,7 +17,7 @@ class Migration(migrations.Migration):
             fields=[
                 ("id", models.AutoField(auto_created=True, primary_key=True, serialize=False, verbose_name="ID")),
-                ("hash_ip_address", models.CharField(db_index=True, max_length=40)),
+                ("hash_ip_address", models.CharField(db_index=True, max_length=64)),
```
Petite remarque sur les migrations : vu que le fichier a été _modifié_, j'imagine que tu as supprimé le fichier d'origine, restauré ta base, et refait une migration ?

Si oui, ce n'est pas la bonne pratique. Accumuler des migrations, ce n'est _pas grave du tout_, et c'est plutôt ça qu'il faut faire. Si vraiment tu en as fait plein, Django a un outil pour (par exemple avant d'intégrer la PR) fusionner plusieurs migrations entre elles. Mais avoir trois, quatre, six migrations pour un même ajout, c'est vraiment pas un drame, d'autant plus si elles sont nommés proprement (et encore).

Le problème que modifier les migrations cause, c'est qu'en l'état, une personne qui a déjà travaillé sur la PR en local (p.ex. pour QA, ou voir, ou que sais-je) devra forcément restaurer sa base à l'état d'avant sa PR (si iel a une sauvegarde), pour relancer les migrations. Ou modifier à la main, ce qui ne devrait jamais être fait. Avoir une migration supplémentaire permet de simplement réappliquer les migrations, sans se poser de question.

PetitMote replied:
Mais c'est Situphen qui m'a demandé de faire comme ça :(

Je comprends la logique de ce que tu dis, mais en même temps ça me semble tellement trop, une migration juste pour corriger une faute de taille de valeur en BDD ><

AmauryCarrade replied:
Ça ne cause vraiment pas de problème d'avoir plein de migrations pour “rien” : le code de migration de Django est de toute façon optimisé pour être très efficient. Je ne suis pas vraiment d'accord avec @Situphen sur ce coup là ^^ .

Si vraiment ça embête, [on peut _squash_ (fusionner) plusieurs migrations entre elles](https://docs.djangoproject.com/fr/2.2/topics/migrations/#squashing-migrations) avec [la commande `squashmigrations`](https://docs.djangoproject.com/fr/2.2/ref/django-admin/#django-admin-squashmigrations). Le système de Django à ce sujet est bien foutu.
